dpdata/abacus/scf.py

```diff
@@ -400,24 +691,61 @@ def make_unlabeled_stru(
         out += "\n"
     out += "\n"
 
+    # ATOMIC_POSITIONS block
     out += "ATOMIC_POSITIONS\n"
     out += "Cartesian    # Cartesian(Unit is LATTICE_CONSTANT)\n"
     # ret += "\n"
-    natom_tot = 0
+    natom_tot = 0  # in for loop, it is also the atom index
     for iele in range(len(data["atom_names"])):
+        if data["atom_numbs"][iele] == 0:
+            continue
         out += data["atom_names"][iele] + "\n"
         out += "0.0\n"
         out += str(data["atom_numbs"][iele]) + "\n"
         for iatom in range(data["atom_numbs"][iele]):
             iatomtype = np.nonzero(data["atom_types"] == iele)[0][iatom]
-            out += "%.12f %.12f %.12f %d %d %d\n" % (
+            iout = "%.12f %.12f %.12f" % ( # 12 decimal places
                 data["coords"][frame_idx][iatomtype, 0],
                 data["coords"][frame_idx][iatomtype, 1],
                 data["coords"][frame_idx][iatomtype, 2],
-                1,
-                1,
-                1,
             )
+            # add flags for move, velocity, mag, angle1, angle2, and sc
+            if move is not None:
+                if isinstance(ndarray2list(move[natom_tot]), (list,tuple)) and len(move[natom_tot]) == 3:
+                    iout += " " + " ".join(["1" if ii else "0" for ii in move[natom_tot]])
+                elif isinstance(ndarray2list(move[natom_tot]), (int,float,bool)):
+                    iout += " " + "1 1 1" if move[natom_tot] else "0 0 0"
+            else:
+                iout += " 1 1 1"
+            
+            if velocity is not None and isinstance(ndarray2list(velocity[natom_tot]), (list,tuple)) and len(velocity[natom_tot]) == 3:
+                iout += " v " + " ".join([f"{ii:.5f}" for ii in velocity[natom_tot]])
```

Comment:
it may not be a good idea to only write 5 digits after decimal in the input file due to the loss of precision.